### ISSUE TITLE
Compare with 2D linear interpolation

### ISSUE THREAD

**Issue by zonca** [zonca/super_res_pysm#7]
@javierhndev use scipy interp2d or similar

**Comment by javierhndev** [comment]
Here it is the spectrum from 4 different images

<img width="586" height="461" alt="Image" src="https://github.com/user-attachments/assets/da17b22d-f527-4a00-b9cf-42b06cc5dce7" />

- **HR**: It is the 256x256 ground truth (the image generated by PySM)
- **LR(64x64)**: It is the reduced image 64x64 (low-res from the HR image)
- **LR(bicubic)**: It its the reduced image but with cubic interpolation to 256x256
- **SR**: It is the result from inference (256x256)

Analyzing these results, the SR shows very similar spectrum to the original HR. And, as we expected, the spectrum from a bicubic interpolation (LR bicubic) is not as good.

However the most obvious issue here it is the LR(64x64) spectrum. It is completly different. I think the reason is due to pixell's `wcs` object. To generate the spectrum for every curve in the plot I generated their pixell's map using the **same** `wcs` obejct. That `wcs` object was saved on a file when I created the 256x256 image dataset. And I think that using the same `wcs` for the 64x64 images is wrong (I'm not that familiar with what is inside the `wcs`)

I think I should create another `wcs` for 64x64 images but maybe you have a better idea. Let me know what you think @zonca

**Comment by zonca** [comment]
@javierhndev correct, you need a different wcs. WCS also knows about pixel size, which is different in the 2 cases.
To create the 64x64 you can use for example:

    coarser = enmap.downgrade(src, 2)   # combine 2x2 pixels → 1

this creates a new wcs at the correct resolution

**Comment by javierhndev** [comment]
Thanks @zonca . I got the *small* `wcs` and now the spectrum looks more consistent.

<img width="586" height="461" alt="Image" src="https://github.com/user-attachments/assets/a34baa0f-81ff-4eab-8109-7f36220ec754" />

**Comment by zonca** [comment]
are you also smoothing the edges before taking power spectrum, right?

**Comment by javierhndev** [comment]
No, I haven't smoothed the edges.  Basically I call this function once I create the map from the Image.
```python
def get_cl_from_pixell_map(imap,lmax):
    # get an apodized mask, multiply the map before doing map2alm
    taper_mask = enmap.apod(pixell.enmap.ones(imap.shape, imap.wcs), width=100)
    alm_taper = curvedsky.map2alm(taper_mask * imap, lmax=lmax)

    # get the correction factor that accounts for the power lost due to only observing a
    # fraction of the sky
    # enmap.pixsizemap is a map of all the physical pixel areas in steradians
    w2 = np.sum(taper_mask.pixsizemap() * taper_mask**2) / (4*np.pi)

    # squaring and averaging over m is done by the alm2cl function
    cl=curvedsky.alm2cl(alm_taper)/w2
    return cl
```

**Comment by zonca** [comment]
ok, good, apodize means smoothing the edges

**Comment by zonca** [comment]
you can also compare with `smoothing` functions in spherical harmonics space

**Comment by zonca** [comment]
the right way of downgrading a map, is going to harmonics space map2alm, then smooth the alm, then alm2map at the lower resolution

**Comment by javierhndev** [comment]
The LR (orange line) differs from the others in the intermediate region

**Comment by zonca** [comment]
`hp.gauss_beam`

**Comment by javierhndev** [comment]
@zonca I've been think about this issue (reducing resolution). We discussed about doing `map2alm` and then smoothing in spherical coordinates. Do you know any methods we can use?

**Comment by zonca** [comment]
in spherical harmonics domain you can smooth using `hp.smooth_alm` or you can use pixell https://github.com/simonsobs/pixell/blob/9dfbf1e88efe30eaa83cbf8f1798a905363dcfd2/pixell/enmap.py#L1429

**Comment by javierhndev** [comment]
I've been using the `pixell.enmap.smooth_gauss(map,sigma)`. I can get smoothed maps but I don't know how to choose the value of `sigma` . This value is basically what specifies the resolution.

**Comment by zonca** [comment]
actually `pixell` has a dedicated tool for this, try 

```python
# imap: 256x256 enmap
factor = 4
imap_64 = enmap.downgrade_fft(imap, factor)
```

**Comment by javierhndev** [comment]
It looks like the `downgrade_fft` is not available fro the pixell version I'm using (0.26). I'm going to check my virtual environment and see if I can upgrade it. It does include the `downgrade` method which gives the same results as what I did in Pytorch (the yellow line in the last plot in this issue)

**Comment by javierhndev** [comment]
I got the same result with `downgrade_fft`'.

But I noticed that the gap reduces when I decrease `width` in  `enmap.apod(pixell.enmap.ones(imap.shape, imap.wcs), width=100)`. Do you know what `width` is? The documentation doesn't explain anything.

**Comment by zonca** [comment]
You can look at this tutorial for the CMB summer school
https://github.com/jeffmcm1977/CMBAnalysis_SummerSchool/blob/6bbbde14c853a938ed42b4f5770e501c5c120eba/CMB_School_Part_04.ipynb

**Comment by javierhndev** [comment]
I finally was able to fix that issues with the different spectrum. Now it looks much better:

<img width="586" height="461" alt="Image" src="https://github.com/user-attachments/assets/bb4d8fc0-bc8e-4676-abe8-e468d2500022" />

I created a pull request (#12)  with the notebook to reproduce this figure. Also, I'm ksaving the config files to run the simulation.

**Comment by zonca** [comment]
I am still a bit concerned about the left part of the plot there. The lines should be one on top of the other. 

I think it is useful to show the code and the plot to chatgpt and see if it recommends any correction that are missing

**Comment by javierhndev** [comment]
I asked Chatgpt (inside vscode) and it told me what I suspected:

> if the SR (AI) image (img_sr) was generated by a neural network, it may not preserve the exact pixel grid, orientation, or WCS alignment as the original HR image, even if the output size matches.
If the pixel grid or WCS is even slightly misaligned, the spherical harmonic transform (map2alm) and resulting cl spectrum will be affected, especially at low l (large angular scales), where the spectrum is most sensitive to global structure and alignment.

>Possible sources of error:
-The SR (AI) image may have been saved or processed with a different origin, orientation, or padding.
-The WCS used for the SR (AI) image may not match the HR image, but the code assumes it does.

I used exactly the same `wcs` for them: hr ,sr(bicubic), sr(ai). That `wcs` was actually the one used to generate the training data.

I also compared the images/results:

<img width="1213" height="418" alt="Image" src="https://github.com/user-attachments/assets/6302bf03-3e2d-4ed7-8ac8-3c64b95adbdf" />

At naked eye it seems good.

The differnece in low *l* could be due some larger structure. So I calculated the (absolute) difference between the HR reference and the super-resolution (both AI and bicubic) and zoomed in:

<img width="824" height="418" alt="Image" src="https://github.com/user-attachments/assets/4ef56183-f2f7-43b0-867e-ad0453592936" />

I don't see anything that could impact the spectrum.... I'll keep thinking about this.

**Comment by zonca** [comment]
can you print out the values of the spectra for ell=1 for all the curves? I don't understand from the graph as they overlap

**Comment by zonca** [comment]
can you also prepare a standalone notebook that I can run myself? It does not need to have the AI run in it, even just reading the output and inspect is fine.

**Comment by javierhndev** [comment]
> can you print out the values of the spectra for ell=1 for all the curves? I don't understand from the graph as they overlap

The ouput of `print(HR[0],SR_AI[0],SR_bic[0],LR[0])` is:
53.61 67.68 53.62 53.85
So only SR_AI doesn't overlap at low *l*.

> can you also prepare a standalone notebook that I can run myself? It does not need to have the AI run in it, even just reading the output and inspect is fine.

In the pull request #12, in the `\analysis_results` folder, there is a notebook with this analysis. You only need the figures from `\analysis_results\results_256_set3_b`

**Comment by zonca** [comment]
when you apodize you need first to remove the mean of the map, so add:

```
    imap = imap - np.mean(imap) 
```

after this the lines agree at low ell:

<img width="586" height="461" alt="Image" src="https://github.com/user-attachments/assets/1c43a162-d499-415b-bbf7-181344531a62" />

it looks like a CMB spectrum. I dont remember, what maps are you using as input?

**Comment by javierhndev** [comment]
Thanks! Removing the mean improved the plot visualization.

> it looks like a CMB spectrum. I dont remember, what maps are you using as input?

I used this:
`
dust = pysm3.ModifiedBlackBodyRealization(
        nside=nside,
        amplitude_modulation_temp_alm="dust_gnilc/raw/gnilc_dust_temperature_modulation_alms_lmax768.fits.gz",
        amplitude_modulation_pol_alm="dust_gnilc/raw/gnilc_dust_polarization_modulation_alms_lmax768.fits.gz",
        largescale_alm="dust_gnilc/raw/gnilc_dust_largescale_template_logpoltens_alm_nside2048_lmax1024_complex64.fits.gz",
        small_scale_cl="dust_gnilc/raw/gnilc_dust_small_scales_logpoltens_cl_lmax16384.fits.gz",
        largescale_alm_mbb_index="dust_gnilc/raw/gnilc_dust_largescale_template_beta_alm_nside2048_lmax1024.fits.gz",
        small_scale_cl_mbb_index="dust_gnilc/raw/gnilc_dust_small_scales_beta_cl_lmax16384_2023.06.06.fits.gz",
        largescale_alm_mbb_temperature="dust_gnilc/raw/gnilc_dust_largescale_template_Td_alm_nside2048_lmax1024.fits.gz",
        small_scale_cl_mbb_temperature="dust_gnilc/raw/gnilc_dust_small_scales_Td_cl_lmax16384_2023.06.06.fits.gz",
        freq_ref="353 GHz",
        seeds=[seed, seed + 1000, seed + 2000],
        max_nside=8192,
)`

    sky = pysm3.Sky(nside=nside, component_objects=[dust])
    freq = 353 * u.GHz
    m = sky.get_emission(freq)
